cartography/intel/aws/ssm.py

```diff
@@ -0,0 +1,192 @@
+        ii["LastPingDateTime"] = dict_date_to_epoch(ii, "LastPingDateTime")
+        ii["RegistrationDate"] = dict_date_to_epoch(ii, "RegistrationDate")
+        ii["LastAssociationExecutionDate"] = dict_date_to_epoch(ii, "LastAssociationExecutionDate")
+        ii["LastSuccessfulAssociationExecutionDate"] = dict_date_to_epoch(ii, "LastSuccessfulAssociationExecutionDate")
+
+    neo4j_session.run(
+        ingest_query,
+        InstanceInformation=data,
+        Region=region,
+        AWS_ACCOUNT_ID=current_aws_account_id,
+        aws_update_tag=aws_update_tag,
+    )
+
+
+@timeit
+def load_instance_patches(
+    neo4j_session: neo4j.Session,
+    data: List[Dict[str, Any]],
+    region: str,
+    current_aws_account_id: str,
+    aws_update_tag: int,
+) -> None:
+    ingest_query = """
+    UNWIND {InstancePatch} AS patch
+        MERGE (p:SSMInstancePatch{id: patch._instance_id + "-" + patch.Title})
+        ON CREATE SET p.firstseen = timestamp()
+        SET p.instance_id = patch._instance_id,
+            p.title = patch.Title,
+            p.kb_id = patch.KBId,
+            p.classification = patch.Classification,
+            p.severity = patch.Severity,
+            p.state = patch.State,
+            p.installed_time = patch.InstalledTime,
+            p.cve_ids = patch.CVEIds,
+            p.region = {Region},
+            p.lastupdated = {aws_update_tag}
+        WITH p
+        MATCH (owner:AWSAccount{id: {AWS_ACCOUNT_ID}})
+        MERGE (owner)-[r:RESOURCE]->(p)
+        ON CREATE SET r.firstseen = timestamp()
+        SET r.lastupdated = {aws_update_tag}
+        WITH p
+        MATCH (owner:AWSAccount{id: {AWS_ACCOUNT_ID}})-[:RESOURCE]->(ec2_instance:EC2Instance{id: p.instance_id})
+        MERGE (ec2_instance)-[r2:HAS_PATCH]->(p)
+        ON CREATE SET r2.firstseen = timestamp()
+        SET r2.lastupdated = {aws_update_tag}
+    """
+    for p in data:
+        p["InstalledTime"] = dict_date_to_epoch(p, "InstalledTime")
+        # Split the comma separated CVEIds, if they exist, and strip
+        # the empty string from the list if not.
+        p["CVEIds"] = list(filter(None, p.get("CVEIds", "").split(",")))
+
+    neo4j_session.run(
+        ingest_query,
+        InstancePatch=data,
+        Region=region,
+        AWS_ACCOUNT_ID=current_aws_account_id,
+        aws_update_tag=aws_update_tag,
+    )
+
+
+@timeit
+def cleanup_ssm(neo4j_session: neo4j.Session, common_job_parameters: Dict) -> None:
+    run_cleanup_job('aws_import_ssm_cleanup.json', neo4j_session, common_job_parameters)
+
+
+@timeit
+def sync(
+    neo4j_session: neo4j.Session, boto3_session: boto3.session.Session, regions: List[str], current_aws_account_id: str,
+    update_tag: int, common_job_parameters: Dict,
+) -> None:
+    for region in regions:
+        logger.info("Syncing SSM for region '%s' in account '%s'.", region, current_aws_account_id)
+        running_instance_ids = get_running_instance_ids(neo4j_session, region, current_aws_account_id)
```

Comment:
Can we also have patches and information for non-running instances.

Reply:
Yeah, I can switch this to not filter by instance state.